metagraph/src/graph/representation/succinct/boss.cpp

```diff
@@ -2273,6 +2288,172 @@ void call_paths(const BOSS &boss,
+        for (uint64_t j = i; j < path.size() - 1; ++j) {
+            set_bit(near_terminal->data(), path[j], async);
+        }
+    }
+    std::optional<edge_index> anchor_edge;
+    if (merge_near_terminal) {
+        anchor_edge = *merge_edge;
+    }
+    callback(std::move(path), anchor_edge);
+}
+
+/**
+ * Traverses all paths that can be visited starting from #edges and invokes
+ * #callback at the end of each path.
+ * A path ends when there are either no outgoing edges from the current node or if the
+ * first node in a fork was already visited.
+ */
+void call_paths_row_diff(
+        const BOSS &boss,
+        std::vector<edge_index>&& edges,
+        const BOSS::Call<std::vector<edge_index> &&, std::optional<edge_index>> &callback,
+        ThreadPool &thread_pool,
+        uint32_t max_length,
+        sdsl::bit_vector *visited,
+        sdsl::bit_vector *terminal,
+        sdsl::bit_vector *near_terminal,
+        ProgressBar &progress_bar) {
+    assert(visited && terminal && near_terminal);
+
+    std::vector<edge_index> out_edges; // stores all branch nodes along the path
+
+    constexpr bool async = true;
+
+    // keep traversing until we have worked off all branches from the queue
+    while (!edges.empty()) {
+        edge_index edge = edges.back();
+        edges.pop_back();
+
+        if (fetch_bit(visited->data(), edge, async))
+            continue;
+
+        std::vector<TAlphabet> sequence = boss.get_node_seq(edge);
+        uint32_t skip_count = 0; // counts number of dummy k-mers to skip
+        while (skip_count < sequence.size() && sequence[skip_count] == boss.kSentinelCode) {
+            skip_count++;
+        }
+
+        std::vector<edge_index> path;
+        path.reserve(100);
+
+        // traverse simple path until we reach its tail or a fork where the first edge
+        // has already been visited
+        while (!fetch_and_set_bit(visited->data(), edge, async)) {
+            assert(edge > 0);
+            ++progress_bar;
+
+            // visit the edge
+            TAlphabet w = boss.get_W(edge) % boss.alph_size;
+
+            // stop the traversal on dummy sink nodes
+            if (w == boss.kSentinelCode)
+                break;
+
+            // don't add dummy source k-mers to the path
+            if (skip_count == 0) {
+                path.push_back(edge);
+            } else {
+                skip_count--;
+            }
+
+            // make one traversal step
+            edge = boss.fwd(edge, w);
+
+            out_edges.resize(0);
+            masked_call_outgoing(boss, edge, nullptr,
+                                 [&](edge_index e) { out_edges.push_back(e); });
+
+            // loop over the outgoing edges (except the first)
+            for (uint32_t i = 1; i < out_edges.size(); ++i) {
+                if (!fetch_bit(visited->data(), out_edges[i], async)) {
+                    edges.push_back(out_edges[i]); // save the edge for visiting later
+                }
```

Comment:
You can just skip these, the other outgoing nodes will be processed with all forks (lines 2681 - ...).

This should reduce the size of the queue a lot. Essentially, you don't need this queue `edges`.

Reply:
Interesting! I assume you and Harun tried this already on the normal `call_paths` and you obtained a speed improvement.

Reply:
I didn't think about this trick before, because `call_paths`, caches the starting k-mer and keep it in the queue to avoid calls `get_node_sequence()`, so we can't use this trick. But here you don't store the last k-mer anyway, so there are no benefits from having the queue. Without it, this will use much less RAM.